packages/versioning/src/versioning.ts

```diff
@@ -528,15 +545,81 @@ export function getVersions(p: Program, t: Type): [Namespace, VersionMap] | [] {
 // these decorators take a `versionSource` parameter because not all types can walk up to
 // the containing namespace. Model properties, for example.
 export function addedAfter(p: Program, type: Type, version: ObjectType) {
+  reportDeprecated(p, "Deprecated: addedAfter is deprecated. Use existsOnVersion instead.", type);
   const appliesAt = appliesAtVersion(getAddedOn, p, type, version);
   return appliesAt === null ? false : !appliesAt;
 }
 
 export function removedOnOrBefore(p: Program, type: Type, version: ObjectType) {
+  reportDeprecated(
+    p,
+    "Deprecated: removedOnOrBefore is deprecated. Use existsOnVersion instead.",
+    type
+  );
   const appliesAt = appliesAtVersion(getRemovedOn, p, type, version);
   return appliesAt === null ? false : appliesAt;
 }
 
+function getSortedVersions(p: Program, t: Type): Version[] | undefined {
+  const [namespace, _] = getVersions(p, t);
+  if (namespace === undefined) return undefined;
+
+  const allVersions = getVersion(p, namespace)?.getVersions();
+  return allVersions?.sort((a, b) => a.index - b.index);
+}
+
+function getFirst(arr: Version[]): Version | undefined {
+  if (!arr.length) return undefined;
+  if (arr.length > 1) {
+    // TODO: Error?
+  }
+  return arr[0];
+}
+
+function getAvailabilityMap(
+  added: Version[],
+  removed: Version[],
+  all: Version[]
+): Map<string, boolean> {
+  const avail = new Map<string, boolean>();
+  // something isn't available by default
+  let isAvail = false;
+  for (const ver of all) {
+    const add = getFirst(added.filter((x) => x.index === ver.index));
+    const rem = getFirst(removed.filter((x) => x.index === ver.index));
+    if (add && rem) {
+      // TODO: Error?
+    }
```

Comment:
Again, this should be part of the validation. It doesn't make sense to say something is added and removed in the same version.